api/models/metadata.go

```diff
@@ -0,0 +1,236 @@
+package models
+
+import (
+	"bytes"
+	"database/sql/driver"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"regexp"
+)
+
+// Metadata encapsulates key-value metadata. The structure is immutable via its public API and nil-safe for its contract
+// permissive nilability is here to simplify updates and reduce the need for nil handling in extensions - metadata should be updated by over-writing the original object:
+//  target.Metadata  = target.Metadata.With("fooKey",1)
+// old MD remains empty
+// Metadata is lenable
+type Metadata map[string]*metadataValue
+
+// metadataValue encapsulates a metadata value in the metadata map,
+// This is stored in its un-parsed JSON format for later (re-) parsing into specific structs or values
+// metadataValue objects are  immutable after JSON load
+type metadataValue []byte
+
+const (
+	maxMetadataValueBytes = 512
+	maxMetadataKeyBytes   = 128
+	maxMetadataKeys       = 100
+)
+
+// Equals is defined based on un-ordered k/v comparison at of the metadata keys and  (compacted) values of metadata, JSON object-value equality for values is property-order dependent
+func (m Metadata) Equals(other Metadata) bool {
+	if len(m) != len(other) {
+		return false
+	}
+	for k1, v1 := range m {
+		v2, _ := other[k1]
+		if !bytes.Equal(*v1, *v2) {
+			return false
+		}
+	}
+	return true
+}
+
+func EmptyMetadata() Metadata {
+	return nil
+}
+
+func (mv *metadataValue) String() string {
+	return string(*mv)
+}
+
+func (v *metadataValue) MarshalJSON() ([]byte, error) {
+	return *v, nil
+}
+
+func (mv *metadataValue) isEmptyValue() bool {
+	sval := string(*mv)
+	return sval == "\"\"" || sval == "null"
+}
+
+// UnmarshalJSON compacts metadata values but does not alter key-ordering for keys
+func (mv *metadataValue) UnmarshalJSON(val []byte) error {
+	buf := bytes.Buffer{}
+	err := json.Compact(&buf, val)
+	if err != nil {
+		return err
+	}
+	if err != nil {
+		return err
+	}
+	*mv = buf.Bytes()
+	return nil
+}
+
+var validKeyRegex = regexp.MustCompile("^[!-~]+$")
```

Comment:
This is not a utf-8 regex. (https://github.com/fnproject/fn/issues/789 says keys are utf-8.)

Reply:
Yep - on feedback from @ericfode  - I dropped this to printable contiguous ascii (no whitesapce)  - it's negociable but being restrictive (initially at least) seemed the safer thing to to do here.